group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java

```diff
@@ -10225,53 +10251,61 @@ public void testConsumerGroupHeartbeatToClassicGroupFromExistingStaticMember() {
+        expectedRecords.add(GroupCoordinatorRecordHelpers.newConsumerGroupMemberSubscriptionRecord(groupId, expectedClassicMember));
+        expectedRecords.add(GroupCoordinatorRecordHelpers.newConsumerGroupEpochRecord(groupId, 0));
+        expectedRecords.add(GroupCoordinatorRecordHelpers.newConsumerGroupTargetAssignmentRecord(groupId, memberId,
+            expectedClassicMember.assignedPartitions()));
+        expectedRecords.add(GroupCoordinatorRecordHelpers.newConsumerGroupTargetAssignmentEpochRecord(groupId, 0));
+        expectedRecords.add(GroupCoordinatorRecordHelpers.newConsumerGroupCurrentAssignmentRecord(groupId, expectedClassicMember));
+
+        // Remove the static member because the rejoining member replaces it.
+        expectedRecords.add(GroupCoordinatorRecordHelpers.newConsumerGroupCurrentAssignmentTombstoneRecord(groupId, memberId));
+        expectedRecords.add(GroupCoordinatorRecordHelpers.newConsumerGroupTargetAssignmentTombstoneRecord(groupId, memberId));
+        expectedRecords.add(GroupCoordinatorRecordHelpers.newConsumerGroupMemberSubscriptionTombstoneRecord(groupId, memberId));
+
+        // Create the new static member.
+        expectedRecords.add(GroupCoordinatorRecordHelpers.newConsumerGroupMemberSubscriptionRecord(groupId, expectedReplacingConsumerMember));
+        expectedRecords.add(GroupCoordinatorRecordHelpers.newConsumerGroupTargetAssignmentRecord(groupId, newMemberId,
+            mkAssignment(mkTopicAssignment(fooTopicId, 0))));
+        expectedRecords.add(GroupCoordinatorRecordHelpers.newConsumerGroupCurrentAssignmentRecord(groupId, expectedReplacingConsumerMember));
+
+        // The static member rejoins the new consumer group.
+        expectedRecords.add(GroupCoordinatorRecordHelpers.newConsumerGroupMemberSubscriptionRecord(groupId, expectedFinalConsumerMember));
+
+        // The subscription metadata hasn't been updated during the conversion, so a new one is computed.
+        expectedRecords.add(GroupCoordinatorRecordHelpers.newConsumerGroupSubscriptionMetadataRecord(groupId,
+            Map.of(fooTopicName, new TopicMetadata(fooTopicId, fooTopicName, 1))));
+
+        // Newly joining static member bumps the group epoch. A new target assignment is computed.
+        expectedRecords.add(GroupCoordinatorRecordHelpers.newConsumerGroupEpochRecord(groupId, 1));
+
+        // If the memberId is generated by the consumer itself, the consumer should retain this memberId.
+        // As a result, the record won't contain a new target assignment record.
+        // If the memberId is not consumer-generated, add a new target assignment record to the expected records,
+        // since a different memberId will be considered as a new member.
+        if (!isConsumerGeneratedMemberId) {
+            expectedRecords.add(GroupCoordinatorRecordHelpers.newConsumerGroupTargetAssignmentRecord(groupId, newMemberId,
+                mkAssignment(mkTopicAssignment(fooTopicId, 0))));
+        }
```

Comment:
Since the consumer's `memberId` won't change, it also means that a new target assignment won't occur.
Therefore, we don't need a new target assignment record here.

Reply:
Hum... I am not sure about this change. My understanding is that we wanted to test a static member rejoining, assuming after a restart so the member id must be different.

Reply:
Got it. 
I will try to simulate this behavior and update the test case accordingly.👍🏼